core_lib/src/managers/selectionmanager.h

```diff
@@ -0,0 +1,131 @@
+#ifndef SELECTIONMANAGER_H
+#define SELECTIONMANAGER_H
+
+#include "pencildef.h"
+#include "basemanager.h"
+#include "movemode.h"
+#include "vertexref.h"
+#include "vectorselection.h"
+
+#include <QPointF>
+#include <QRectF>
+#include <QPolygonF>
+#include <QTransform>
+
+class Editor;
+
+class SelectionManager : public BaseManager
+{
+    Q_OBJECT
+public:
+    explicit SelectionManager(Editor* editor);
+    ~SelectionManager() override;
+
+    bool init() override;
+    Status load(Object*) override;
+    Status save(Object*) override;
+    void workingLayerChanged(Layer*) override;
+
+    QVector<QPointF> calcSelectionCenterPoints();
+
+    void updatePolygons();
+    void updateTransformedSelection() { mTransformedSelection = mTempTransformedSelection; }
+
+    QRectF mappedSelection();
+
+    QPointF whichAnchorPoint(QPointF currentPoint);
+    QPointF getTransformOffset() { return mOffset; }
+    QPointF offsetFromAspectRatio(qreal offsetX, qreal offsetY);
+
+    void flipSelection(bool flipVertical);
+
+    void setSelection(QRectF rect);
+
+    void translate(QPointF point);
+
+    MoveMode getMoveModeForSelectionAnchor(QPointF pos);
+    MoveMode validateMoveMode(QPointF pos);
+    MoveMode getMoveMode() const { return mMoveMode; }
+    void setMoveMode(MoveMode moveMode) { mMoveMode = moveMode; }
+
+    bool somethingSelected() const { return mSomethingSelected; }
+
+    void calculateSelectionTransformation();
+    void adjustSelection(const QPointF& currentPoint, qreal offsetX, qreal offsetY, qreal rotationOffset);
+    MoveMode moveModeForAnchorInRange(QPointF lastPos);
+    void setCurves(QList<int> curves) { mClosestCurves = curves; }
+    void setVertices(QList<VertexRef> vertices) { mClosestVertices = vertices; }
+
+    void clearCurves();
+    void clearVertices();
+
+    const QList<int> closestCurves() { return mClosestCurves; }
+    const QList<VertexRef> closestVertices() { return mClosestVertices; }
+
+    QTransform selectionTransform() { return mSelectionTransform; }
+    void setSelectionTransform(QTransform transform) { mSelectionTransform = transform; }
+    void resetSelectionTransform();
+
+    inline bool transformHasBeenModified() { return (mSelection != mTempTransformedSelection) || mRotatedAngle != mOldRotatedAngle; }
```

Comment:
You're right, I've redone it now. Now it's doing a fuzzy compare comparison against both values in absolute context.

Reply:
I know I said a modulus wasn't necessary, but I've changed my mind 😢 . The documentation for the atan2 function says that it returns the:
> Principal arc tangent of y/x, in the interval [-pi,+pi] radians.

Since that range is inclusive, the angles -pi and +pi can be returned and are equivalent to each other. Also, -angle != angle, so qAbs should not be used (ex. -10º is congruent to 350º not 10º). May I recommend using the following line for `rotationHasBeenModified`:
```
return !qFuzzyCompare(fmod(mRotatedAngle+360,360),fmod(mRotationOffset+360,360));
```
This is still not entirely correct when mRotatedAngle < -360, however that shouldn't happen 🤞 . To fix this you would need to add another fmod around mRotatedAngle and mRotationOffset. I personally feel like that's overkill, but I don't care too much one way or the other.

Reply:
Hmm I feel like it could be done easier, In radians.. what if do
```
qFuzzyCompare(qAbs(mRotatedAngle*M_PI/180),qAbs(mRotationOffset*M_PI/180))
```
Won't that work?

Reply:
No, still the same problem. Let's say you have angles in radians of pi/2 and -pi/2 (90º and -90º respectively). They are not the same angle (the first points "up"/+y axis and second points "down"/-y axis), but the absolute value of those angles is the same.

Reply:
Here's a graphical representation of the difference from x = -2pi to 2pi or equivalently -360 to 360):

y = abs(x):
![Screen Shot 2019-06-24 at 12 57 33 PM](https://user-images.githubusercontent.com/3461051/60044754-4b9be900-9680-11e9-8281-7c62b6db4f5d.png)

y = x % 2pi:
![Screen Shot 2019-06-24 at 1 00 34 PM](https://user-images.githubusercontent.com/3461051/60044760-4f2f7000-9680-11e9-81a1-15e3d0afae9d.png)

Reply:
The reason that we need to add 360 or 2pi is because the modulus function in C++ returns a negative number sometimes, which results in a graph like this:
![Screen Shot 2019-06-24 at 1 06 16 PM](https://user-images.githubusercontent.com/3461051/60044985-ded51e80-9680-11e9-8982-55005b4308eb.png)

Reply:
Technically or mathematically it's not the same but visually the user won't see the difference, so why does it matter? 2 radians will look the same as -2 radians.. you might have turned the rotated the transformation but the image itself won't change so visually you haven't modified the image.